### DISCUSSION TITLE
How to efficiently get a cluster ID given a sample ID of which sample has been added to a IVF index?

### DISCUSSION THREAD

**Discussion by yzou2** [facebookresearch/faiss#3555] [Q&A]
Currently I have an ID of a vector which has been added to a IVF index and would like to get the cluster assignment. A naive way is firstly defining a mapping from the sample ID to cluster ID given the inverted list and then use the mapping to find the cluster assignment. But is there a build-in function to support fast mapping from sample ID to cluster ID without looking to the inverted list?

**Comment by mdouze** [comment]
There is no difference between the cluster ID and the inverted list. 
You can get the cluster IDs with `index.quantizer.search(vector, 1)[1]`.

If I misunderstood the question, feel free to correct.

**Reply by asadoughi** [reply]
Re-phrasing the question as: is there an O(1) way of getting the cluster assignment for a given vector id? Invoking `search()` requires a k-NN computation.

**Reply by mdouze** [reply]
Right, you need a direct map for that see doc here: 
https://github.com/facebookresearch/faiss/wiki/Special-operations-on-indexes#direct-map-for-an-indexivf
and demo here: [issue_3555.ipynb](https://gist.github.com/mdouze/1876557196b72cd71313fc25fca73d77#file-issue_3555-ipynb)
